File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/gui/slot/EcoSlotBuilder.java
package com.willfp.eco.internal.gui.slot;

import com.willfp.eco.core.gui.slot.Slot;
import com.willfp.eco.core.gui.slot.SlotBuilder;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.function.BiConsumer;
import java.util.function.Function;

public class EcoSlotBuilder implements SlotBuilder {
    private final Function<Player, ItemStack> provider;

    private BiConsumer<InventoryClickEvent, Slot> onLeftClick = null;

    private BiConsumer<InventoryClickEvent, Slot> onRightClick = null;

    private BiConsumer<InventoryClickEvent, Slot> onShiftLeftClick = null;

    private BiConsumer<InventoryClickEvent, Slot> onShiftRightClick = null;

    private BiConsumer<InventoryClickEvent, Slot> onMiddleClick = null;

    public EcoSlotBuilder(@NotNull final Function<Player, ItemStack> provider) {
        this.provider = provider;
    }

    public SlotBuilder onLeftClick(@NotNull final BiConsumer<InventoryClickEvent, Slot> action) {
        this.onLeftClick = action;
        return this;
    }

    public SlotBuilder onRightClick(@NotNull final BiConsumer<InventoryClickEvent, Slot> action) {
        this.onRightClick = action;
        return this;
    }

    public SlotBuilder onShiftLeftClick(@NotNull final BiConsumer<InventoryClickEvent, Slot> action) {
        this.onShiftLeftClick = action;
        return this;
    }

    public SlotBuilder onShiftRightClick(@NotNull final BiConsumer<InventoryClickEvent, Slot> action) {
        this.onShiftRightClick = action;
        return this;
    }

    public SlotBuilder onMiddleClick(@NotNull final BiConsumer<InventoryClickEvent, Slot> action) {
        this.onMiddleClick = action;
        return this;
    }

    public Slot build() {
        return new EcoSlot(provider, onLeftClick, onRightClick, onShiftLeftClick, onShiftRightClick, onMiddleClick);
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/gui/slot/EcoSlotBuilder.kt
package com.willfp.eco.internal.gui.slot

import com.willfp.eco.core.gui.slot.Slot
import com.willfp.eco.core.gui.slot.SlotBuilder
import com.willfp.eco.core.gui.slot.functional.CaptiveFilter
import com.willfp.eco.core.gui.slot.functional.SlotHandler
import com.willfp.eco.core.gui.slot.functional.SlotProvider
import com.willfp.eco.core.gui.slot.functional.SlotUpdater
import com.willfp.eco.core.map.listMap
import org.bukkit.entity.Player
import org.bukkit.event.inventory.ClickType
import java.util.function.Predicate

class EcoSlotBuilder(private val provider: SlotProvider) : SlotBuilder {
    private var captive = false
    private var captiveFromEmpty = false
    private var updater: SlotUpdater = SlotUpdater { player, menu, _ -> provider.provide(player, menu) }

    private val handlers = listMap<ClickType, SlotHandler>()

    private var captiveFilter =
        CaptiveFilter { _, _, _ -> true }
    private var notCaptiveFor: (Player) -> Boolean = { _ -> false}

    override fun onClick(type: ClickType, action: SlotHandler): SlotBuilder {
        handlers[type].add(action)
        return this
    }

    override fun notCaptiveFor(predicate: Predicate<Player>): SlotBuilder {
        notCaptiveFor = { player -> predicate.test(player) }
        return this
    }

    override fun setCaptiveFilter(filter: CaptiveFilter): SlotBuilder {
        captiveFilter = filter
        return this
    }

    override fun setCaptive(fromEmpty: Boolean): SlotBuilder {
        captive = true
        captiveFromEmpty = fromEmpty
        return this
    }

    override fun setUpdater(updater: SlotUpdater): SlotBuilder {
        this.updater = updater
        return this
    }

    override fun build(): Slot {
        return if (captive) {
            EcoCaptiveSlot(
                provider,
                captiveFromEmpty,
                notCaptiveFor,
                captiveFilter
            )
        } else {
            EcoSlot(
                provider,
                handlers,
                updater
            )
        }
    }
}
